### PR TITLE
feat: request options

### LINKED ISSUE #392: Enable set ky's options to Bee instance and override them per call

### PR DESCRIPTION
**Pull request by AuHau**
Closes #392 and also allows setting custom `fetch` instance which will be needed for the upload progress support.

### CODE REVIEW

File: src/bee.ts [hunk 1]
```diff
@@ -961,27 +1003,27 @@ export class Bee {
...
   /**
    * Ping the Bee node to see if there is a live Bee node on the given URL.
    *
    * @returns true if successful, false on error
    */
-  async isConnected(): Promise<boolean> {
+  async isConnected(options?: RequestOptions): Promise<boolean> {
     try {
-      await status.checkConnection(this.ky)
+      await status.checkConnection(this.getKy(options))
```
Comment (nugaon) [inline src/bee.ts:1026]: you don't assert `RequestOptions`

File: src/bee.ts [hunk 2]
```diff
@@ -913,6 +954,7 @@ export class Bee {
    * @deprecated Use DebugBee for postage batch management
    */
   async createPostageBatch(amount: NumberString, depth: number, options?: PostageBatchOptions): Promise<BatchId> {
+    assertRequestOptions(options, 'PostageBatchOptions')
```
Comment (nugaon) [inline src/bee.ts:957]: it is a bit strange you assert with an other - although parent - type's assert method and you check its other properties later. why don't you make one for `PostageBatchOptions`?

File: src/bee-debug.ts [hunk 3]
```diff
@@ -257,6 +288,7 @@ export class BeeDebug {
    * @param options.gasLimit Gas limit for the cashout transaction in WEI
    */
   async cashoutLastCheque(address: string | Address, options?: CashoutOptions): Promise<string> {
+    assertRequestOptions(options)
```
Comment (nugaon) [inline src/bee-debug.ts:291]: same as at `PostageBatchOptions`

File: src/bee.ts [hunk 4]
```diff
@@ -961,27 +1003,27 @@ export class Bee {
...
 
   /**
    * Ping the Bee node to see if there is a live Bee node on the given URL.
    *
    * @throws If connection was not successful throw error
    */
-  async checkConnection(): Promise<void> | never {
-    return status.checkConnection(this.ky)
+  async checkConnection(options?: RequestOptions): Promise<void> | never {
+    return status.checkConnection(this.getKy(options))
```
Comment (nugaon) [inline src/bee.ts:1016]: you don't assert `RequestOptions`

File: src/bee.ts [hunk 5]
```diff
@@ -961,27 +1003,27 @@ export class Bee {
    * @see [Bee API reference - `GET /stamps`](https://docs.ethswarm.org/api/#tag/Postage-Stamps/paths/~1stamps/get)
    * @deprecated Use DebugBee for postage batch management
    */
-  async getAllPostageBatch(): Promise<PostageBatch[]> {
-    return stamps.getAllPostageBatches(this.ky)
+  async getAllPostageBatch(options?: RequestOptions): Promise<PostageBatch[]> {
+    return stamps.getAllPostageBatches(this.getKy(options))
```
Comment (nugaon) [inline src/bee.ts:1007]: you don't assert `RequestOptions`

File: src/bee.ts [hunk 6]
```diff
@@ -946,10 +988,10 @@ export class Bee {
    * @see [Bee API reference - `GET /stamps/${id}`](https://docs.ethswarm.org/api/#tag/Postage-Stamps/paths/~1stamps~1{id}/get)
    * @deprecated Use DebugBee for postage batch management
    */
-  async getPostageBatch(postageBatchId: BatchId | string): Promise<PostageBatch> {
+  async getPostageBatch(postageBatchId: BatchId | string, options?: RequestOptions): Promise<PostageBatch> {
     assertBatchId(postageBatchId)
 
-    return stamps.getPostageBatch(this.ky, postageBatchId)
+    return stamps.getPostageBatch(this.getKy(options), postageBatchId)
```
Comment (nugaon) [inline src/bee.ts:994]: you don't assert `RequestOptions`

**Review by nugaon — COMMENTED** [review]
I like how the testing is structured, but I'm really missing the upload progress test in puppeteer browser environment where this `extend` functionality of `ky` and this desired behaviour could be tested. 
Currently, only the asserting is checked, but not the functionality.

**Comment by AuHau** [comment]
Yeah, you are right. I will extend the tests to test the `extend` part, but it won't be on the upload progress. That one won't be covered in tests as it requires another dependency that even in its current form is not compatible with this approach (my PR that tweaks the package still has not been merged). I plan to have an example that will cover this use case.

**Comment by AuHau** [comment]
@nugaon waiting for your approval and then will rebase.

File: src/bee-debug.ts [hunk 7]
```diff
@@ -257,17 +289,11 @@ export class BeeDebug {
...
```
Comment (nugaon) [inline src/bee-debug.ts:292]: you are asserting this in `assertCashoutOptions`

**Review by nugaon — APPROVED** [review]
> I will extend the tests to test the `extend` part, but it won't be on the upload progress. That one won't be covered in tests as it requires another dependency that even in its current form is not compatible with this approach (my PR that tweaks the package still has not been merged). I plan to have an example that will cover this use case.

I don't completely get this, please before merge ping me to discuss this.
for test cases you can add packages to `devDependencies` in a justified case (like now).

please correct that part that I just commented now

File: test/unit/assertions.ts [hunk 8]
```diff
@@ -1,6 +1,7 @@
 /* eslint-disable */
-import { BeeArgumentError } from '../../src'
+import { BeeArgumentError, BeeOptions } from '../../src'
 import { makeBytes } from '../../src/utils/bytes'
+import { assertCashoutOptions } from '../../src/utils/type'
```
Comment (nugaon) [inline test/unit/assertions.ts:4]: unused import

**Review by nugaon — APPROVED** [review]
LGTM!

before release please create an example for using altered fetch instance that can track upload process

**Pull request merged**